lib/omniauth/auth0/jwt_validator.rb

```diff
@@ -24,27 +28,56 @@ def initialize(options)
 
         @client_id = options.client_id
         @client_secret = options.client_secret
+
+        @leeway = authorize_params[:leeway] || 60
+        @max_age = authorize_params[:max_age]
+        @nonce = authorize_params[:nonce]
+      end
+
+      def verify_signature(jwt)
+        head = token_head(jwt)
+
+        # Make sure the algorithm is supported and get the decode key.
+        if head[:alg] == 'RS256'
+          [rs256_decode_key(head[:kid]), head[:alg]]
+        elsif head[:alg] == 'HS256'
+          [@client_secret, head[:alg]]
+        else
+          raise OmniAuth::Auth0::TokenValidationError.new("Signature algorithm of #{head[:alg]} is not supported. Expected the ID token to be signed with RS256 or HS256")
+        end
       end
 
       # Decode a JWT.
       # @param jwt string - JWT to decode.
       # @return hash - The decoded token, if there were no exceptions.
       # @see https://github.com/jwt/ruby-jwt
       def decode(jwt)
-        head = token_head(jwt)
+        if !jwt
+          raise OmniAuth::Auth0::TokenValidationError.new('ID token is required but missing')
+        end
 
-        # Make sure the algorithm is supported and get the decode key.
-        decode_key = @client_secret
-        if head[:alg] == 'RS256'
-          decode_key = rs256_decode_key(head[:kid])
-        elsif head[:alg] != 'HS256'
-          raise JWT::VerificationError, :id_token_alg_unsupported
+        parts = jwt.split('.')
+        if parts.length != 3
+          raise OmniAuth::Auth0::TokenValidationError.new('ID token could not be decoded')
         end
 
+        key, alg = verify_signature(jwt)
+
         # Docs: https://github.com/jwt/ruby-jwt#algorithms-and-usage
-        JWT.decode(jwt, decode_key, true, decode_opts(head[:alg]))
+        id_token, header = JWT.decode(jwt, key, false)
```

Comment:
For context `It is strongly recommended that you hard code the algorithm, as you may leave yourself vulnerable by dynamically picking the algorithm`

This would be a concern if we weren't verifying the algorithm in `verify_signature` https://github.com/auth0/omniauth-auth0/pull/92/files#diff-fe7c85400c4abddbb14f08149fd91c58R37

Basically what we've done here is reduce the use of `ruby-jwt` to decoding the keys, removed the verification process and implemented it here in this library for more coverage and consistency across all our SDKs

Reply:
Might be good to leave a comment to that extent, as someone looking at the code might see the recommended usage and have the same concern.